### ISSUE TITLE
Would this allow me to work with Go?

### ISSUE THREAD

**Issue by phumberdroz** [ZOTTCE/fivem-wasm#3]
So I have never worked with WASM but I am working a lot with Go in the past. It seems to be that enabling this WASM runtime should allow me to run Go on the Client as well as on the server is this assumption correct?

**Comment by zottce** [comment]
yes if someone will create go library to work with the runtime

**Comment by phumberdroz** [comment]
So I tried building your example but it is failing due to
```
   Compiling cfx-server v0.2.0 (C:\Users\phumb\projects\gta\fivem-wasm\bindings\server)
   Compiling cfx-client v0.2.0 (C:\Users\phumb\projects\gta\fivem-wasm\bindings\client)
error[E0583]: file not found for module `natives`
 --> bindings\client\src\lib.rs:2:1
  |
2 | pub mod natives;
  | ^^^^^^^^^^^^^^^^
  |
  = help: to create the module `natives`, create file "bindings\client\src\natives.rs"

error[E0583]: file not found for module `natives`
 --> bindings\server\src\lib.rs:3:1
  |
3 | pub mod natives;
  | ^^^^^^^^^^^^^^^^
  |
  = help: to create the module `natives`, create file "bindings\server\src\natives.rs"

error[E0432]: unresolved imports `crate::natives::player`, `crate::natives::streaming`, `crate::natives::task`
 --> bindings\client\src\task.rs:1:22
  |
1 | use crate::natives::{player::*, streaming::*, task::*};
  |                      ^^^^^^     ^^^^^^^^^     ^^^^ could not find `task` in `natives`
  |                      |          |
  |                      |          could not find `streaming` in `natives`
  |                      could not find `player` in `natives`
```
But I might be missing something. 

C:\Users\phumb\projects\gta\fivem-wasm>cargo wasi build --release --all-features

**Comment by Roms1383** [comment]
hey, I also just gave a try at building which yields the same error as @phumberdroz mentioned :

<img width="499" alt="Screen Shot 2564-07-16 at 15 09 32" src="https://user-images.githubusercontent.com/21016014/125916211-b78f90bf-a901-4666-a697-d504ffd458ec.png">
<img width="508" alt="Screen Shot 2564-07-16 at 15 09 45" src="https://user-images.githubusercontent.com/21016014/125916250-b427e9ca-9ccd-459b-958f-2919f4ddba5a.png">
<img width="172" alt="Screen Shot 2564-07-16 at 15 09 56" src="https://user-images.githubusercontent.com/21016014/125916303-9ba9e383-4782-44bc-9101-4e0e07852cf4.png">
<img width="572" alt="Screen Shot 2564-07-16 at 15 10 06" src="https://user-images.githubusercontent.com/21016014/125916339-f8f69997-f804-43d8-86f8-3393fb2c6d57.png">
<img width="571" alt="Screen Shot 2564-07-16 at 15 14 43" src="https://user-images.githubusercontent.com/21016014/125916418-cda98ac5-8fdc-4aeb-a11c-3b1211d7db7d.png">
<img width="279" alt="Screen Shot 2564-07-16 at 15 16 11" src="https://user-images.githubusercontent.com/21016014/125916431-48890c0a-fb8d-4d27-a3d0-c1d6b79a4358.png">
<img width="786" alt="Screen Shot 2564-07-16 at 15 17 14" src="https://user-images.githubusercontent.com/21016014/125916443-c40ddaf7-1cca-4ae0-a2cd-1eb039c372fd.png">

I guess I'm probably missing something somewhere, could you provide a bit more information on how to use your crate ?

Thanks for the initiative, I'm personally kinda hyped that somebody takes the time to provide support for Rust and Wasm to the FiveM / RedM ecosystem 👍

**Comment by zottce** [comment]
You use code from the `master` branch, my code is living on `wasm` (for https://github.com/zottce/fivem).

Also, I do not ship `natives.rs` files in this repo until citizenfx/fivem#744 isn't merged. It will require some work to generate bindings to native functions.

For now, you can use [this crate](https://github.com/ZOTTCE/fivem-wasm/tree/master/natives-gen) to do that (but make sure you changed hard-coded links in the `main.rs` file).

It's the outdated building guide btw. You do not need to clone this repo into FiveM source code. They are separated. Just build FiveM as their docs say and then you can build these bindings.

**Comment by Roms1383** [comment]
Well in the meantime I realized I had to generate the `natives-gen` (and yes, I changed the links in `main.rs`) but I ran into another issue, even when switching to `wasm` branch : the script cannot find the [inp/natives_cfx.lua](https://github.com/ZOTTCE/fivem-wasm/blob/e2617f557b5a3bbf9afb8284921de4e8bfa5e1ee/natives-gen/src/main.rs#L19) and [inp/natives_global.lua](https://github.com/ZOTTCE/fivem-wasm/blob/e2617f557b5a3bbf9afb8284921de4e8bfa5e1ee/natives-gen/src/main.rs#L24).

**Comment by Roms1383** [comment]
Not sure if I'll be able to process any further since [FiveM provided build documentation](https://github.com/citizenfx/fivem/blob/master/docs/building.md) is for Windows while I'm on a MacOS :thinking:

**Comment by zottce** [comment]
> 
> 
> Well in the meantime I realized I had to generate the `natives-gen` (and yes, I changed the links in `main.rs`) but I ran into another issue, even when switching to `wasm` branch : the script cannot find the [inp/natives_cfx.lua](https://github.com/ZOTTCE/fivem-wasm/blob/e2617f557b5a3bbf9afb8284921de4e8bfa5e1ee/natives-gen/src/main.rs#L19) and [inp/natives_global.lua](https://github.com/ZOTTCE/fivem-wasm/blob/e2617f557b5a3bbf9afb8284921de4e8bfa5e1ee/natives-gen/src/main.rs#L24).

Those two files are created in FiveM pre-build scripts.



> 
> 
> Not sure if I'll be able to process any further since [FiveM provided build documentation](https://github.com/citizenfx/fivem/blob/master/docs/building.md) is for Windows while I'm on a MacOS 🤔

So, to use WASM runtime before it is merged you should manually compile FiveM. Parallels or cloud-based VM is your choice.

**Comment by Roms1383** [comment]
as a side question, is there any relationship with these crates : https://crates.io/search?q=cfx ?

**Comment by zottce** [comment]
> 
> 
> as a side question, is there any relationship with these crates : https://crates.io/search?q=cfx ?

Yeap, they are mine. I thought someone could take these names before I do.
